ucp/_libs/ucx_api.pyx

```diff
@@ -0,0 +1,490 @@
+            ucp_listener_destroy(self.handle)
+
+
+cdef class UCXContext:
+    cdef:
+        ucp_context_h _context
+        bint _initialized
+        object _config
+
+    def __cinit__(self, config_dict):
+        cdef ucp_params_t ucp_params
+        cdef ucp_worker_params_t worker_params
+        cdef ucs_status_t status
+        self._initialized = False
+
+        memset(&ucp_params, 0, sizeof(ucp_params))
+        ucp_params.field_mask = (UCP_PARAM_FIELD_FEATURES |  # noqa
+                                 UCP_PARAM_FIELD_REQUEST_SIZE |  # noqa
+                                 UCP_PARAM_FIELD_REQUEST_INIT)
+
+        # We always request UCP_FEATURE_WAKEUP even when in blocking mode
+        # See <https://github.com/rapidsai/ucx-py/pull/377>
+        ucp_params.features = (UCP_FEATURE_TAG |  # noqa
+                               UCP_FEATURE_WAKEUP |  # noqa
+                               UCP_FEATURE_STREAM)
+
+        ucp_params.request_size = sizeof(ucp_request)
+        ucp_params.request_init = (
+            <ucp_request_init_callback_t>ucp_request_reset
+        )
+
+        cdef ucp_config_t *config = _read_ucx_config(config_dict)
+        status = ucp_init(&ucp_params, config, &self._context)
+        assert_ucs_status(status)
+        self._initialized = True
+
+        self._config = ucx_config_to_dict(config)
+        ucp_config_release(config)
+
+        logging.info("UCP initiated using config: ")
+        for k, v in self._config.items():
+            logging.info("  %s: %s" % (k, v))
+
+    def destroy(self):
+        if self._initialized:
+            self._initialized = False
+            ucp_cleanup(self._context)
+
+    def get_config(self):
+        return self._config
+
+
+cdef class UCXWorker:
+    cdef:
+        ucp_worker_h _handle
+        bint _initialized
+
+    def __cinit__(self, UCXContext context):
+        cdef ucp_params_t ucp_params
+        cdef ucp_worker_params_t worker_params
+        cdef ucs_status_t status
+        self._initialized = False
+        memset(&worker_params, 0, sizeof(worker_params))
+        worker_params.field_mask = UCP_WORKER_PARAM_FIELD_THREAD_MODE
+        worker_params.thread_mode = UCS_THREAD_MODE_MULTI
+        status = ucp_worker_create(context._context, &worker_params, &self._handle)
+        assert_ucs_status(status)
+        self._initialized = True
+
+    def destroy(self):
+        if self._initialized:
+            self._initialized = False
+            ucp_worker_destroy(self._handle)
+
+    def init_blocking_progress_mode(self):
+        # In blocking progress mode, we create an epoll file
+        # descriptor that we can wait on later.
+        cdef ucs_status_t status
+        cdef int ucp_epoll_fd
+        cdef epoll_event ev
+        cdef int err
+        status = ucp_worker_get_efd(self._handle, &ucp_epoll_fd)
+        assert_ucs_status(status)
+        self.arm()
+        epoll_fd = epoll_create(1)
+        assert(epoll_fd != -1)
```

Comment:
Should we raise some sort of `IOError` or other appropriate error here?

Reply:
Fixed